### PR TITLE
Distinguish mandatory and optional dependencies

### PR DESCRIPTION
**Pull request by koraktor**
See #57

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -11,6 +11,9 @@ en:
       deps_html:
         one:   "%{name} requires the following formula to be installed:"
         other: "%{name} requires the following formulae to be installed:"
+      optdeps_html:
```
Comment (koraktor) [inline config/locales/en.yml]: I’m not totally happy with the wording, but this is the best I could come up with.

**Comment by MikeMcQuaid** [comment]
I wonder if it's worth going the full hog and differentiating between required, optional and build dependencies.

**Comment by koraktor** [comment]
Looks like I got something wrong here. I already thought about it while writing the patch.

The new implementation will only show build and optional dependencies – which is obviously wrong. Users would definitely want to see build *and* required dependencies (IMHO there’s no need for additional separation here) and optional dependencies.

Brew‘s `Formula#to_hash` doesn’t have a separate field for required dependencies. Can we just subtract optional dependencies from all dependencies to get build *and* required dependencies?

**Comment by koraktor** [comment]
@MikeMcQuaid I fixed a bunch of other code smells and did some loosely related work on `FormulaeHelper`. Maybe you want to have another look.

**Comment by koraktor** [comment]
I guess I figured this out – 0f523add3627c83bb88bb570678509f32ea8a6bd should do the trick. Any further comments?

File: app/models/formula.rb [hunk 2]
```diff
@@ -74,9 +79,11 @@ def update_metadata(formula_info)
...
     self.revision = formula_info['revision']
 
-    self.deps = formula_info['dependencies'].map do |dep|
-      repository.formulae.find_by(name: dep) || Repository.core.formulae.find_by(name: dep)
-    end
+    optdeps = formula_info['optional_dependencies'] +
+              formula_info['recommended_dependencies']
+    self.optdeps = optdeps.map { |dep| find_dependency dep }
+    deps = formula_info['dependencies'] - optdeps
+    self.deps = deps.map { |dep| find_dependency dep }
```
Comment (MikeMcQuaid) [inline app/models/formula.rb]: I think this is still a little strange; it doesn't differentiate between optional but on-by-default and optional but off-by-default. If I were you I'd have four categories of dependencies:

- recommended/on-by-default optional dependencies (`formula_info['recommended_dependencies']`)
- optional/off-by-default dependencies (`formula_info['optional_dependencies']`)
- build-from-source/build-time dependencies (`formula_info['build_dependencies']`)
- mandatory runtime dependencies (`formula_info['dependencies'] - ` all the above)

**Comment by koraktor** [comment]
@MikeMcQuaid Finally found the time for the next iteration on this one.

**Pull request closed without merging**